Objects/regions.c

```diff
@@ -1337,12 +1429,146 @@ int _Py_is_bridge_object(PyObject *op) {
+        // to the bridge are created. This means that the bridge might have multiple
+        // unknown references.
+        //
+        // Using the object RC is an over approximation, since internal cycles from
+        // objects to the bridge object will also increase the RC thereby tricking
+        // this check into opening it again.
+        if (Py_REFCNT(bridge) > rc_limit) {
+            regionmetadata_open(Py_REGION_DATA(bridge));
+        }
+
+        // If it's closed there is nothing we need to do.
+        if (!regionmetadata_is_open(old_data)) {
+            continue;
+        }
+
+        // Create the new `regionmetadata*`
+        regionmetadata* new_data = (regionmetadata*)calloc(1, sizeof(regionmetadata));
+        if (!new_data) {
+            PyErr_NoMemory();
+            goto fail;
+        }
+        _Py_CAST(PyRegionObject *, bridge)->metadata = new_data;
+        regionmetadata_inc_rc(new_data);
+        new_data->bridge = _Py_CAST(PyRegionObject*, bridge);
+        Py_XSETREF(new_data->name, old_data->name);
+        regionmetadata_open(new_data);
+        regionmetadata_set_parent(new_data, regionmetadata_get_parent(old_data));
+        new_data->cown = old_data->cown;
+        old_data->cown = NULL;
+
+        // Merge the old region data into local. This has to be done after the
+        // created of the `new_data` to prevent the parent from closing
+        // premeturely when the old data gets detached from it.
+        regionmetadata_set_parent(old_data, NULL);
+        regionmetadata_merge(old_data, _Py_LOCAL_REGION);
+
+        Py_SET_REGION(bridge, new_data);
+        new_data->lrc += Py_REFCNT(bridge);
+        // Only subtract 1 from the LRC if the reference comes from a parent.
+        // Owning references from the local region should still count towards
+        // the LRC.
+        if (regionmetadata_has_parent(new_data) || new_data->cown) {
+            new_data->lrc -= 1;
+        }
+
+        if (stack_push(info.pending, bridge)) {
+            goto fail;
+        }
+
+        // Re-add everything to the current region
+        while (!stack_empty(info.pending)) {
+            PyObject *item = stack_pop(info.pending);
+
+            // Add `info.src` for better error messages
+            info.src = item;
+
+            if (!visit_object(item, (visitproc)_add_to_region_visit, &info)) {
+                goto fail;
+            }
+        }
+
+        // Mark the region as clean
+        regionmetadata_mark_as_not_dirty(new_data);
+
+        // The LRC will never decrease after this point. If the region is open
+        // due to the LRC it will remain open and the close fails.
+        if (new_data->lrc > lrc_limit) {
+            break;
+        }
+
+        // Update the open status and make sure the parent knows
+        if (new_data->osc == 0) {
+            if (regionmetadata_close(new_data) != 0) {
+                goto fail;
```

Comment:
Could this failure cause a leak or an inconsistent state?

Reply:
Added a comment